Add oneof support in codec v4

dfenucci opened the pull request.

tsaubergine commented:
Hi -

Thank you very much for this pull request. A quick look at the code looks very good to me. 
Could you please add/update the following so I can merge this:

- Make sure the unit tests build so that the automated tests complete successfully (currently `dccl_test_oneof` fails).
- Can you extend `dccl_test_oneof` to add a few more tests? I'd like to verify the max/min sizes and perhaps add some `oneof` fields with more than two members. If you have a real message you're using that you can add to the unit test, all the better.
 - Please read and sign the AUTHORS file by adding your name, email and date to the bottom of the file and commit it to this branch. This confirms you're willing to distribute these changes under the existing open source licenses.
- Could you update the documentation (https://github.com/GobySoft/dccl/blob/4.0/src/doc/doxy_body/codecs.dox) to include the oneof encode/decode logic in the algorithm?

Thanks again for this valuable contribution!

dfenucci commented:
Hi Toby,

here are the main updates I've done so far.

- Added my name / contact to the AUTHORS file
- Updated the documentation page with oneof algorithm. Let me know if you'd like me to extend it further.
- I've fixed the `dccl_oneof` unit test in [this commit](https://github.com/GobySoft/dccl/pull/70/commits/555d159d7946b0df893ca129d6624f45a0319b45), although I'm not sure this is the best way to approach the issue. Do you have any better idea?
- I still need to extend the unit test. In this regard I wanted to ask you: does the expected min/max size of a message include the header as well (i.e. the `dccl.msg.id` and the fields with `in_head` set to true)?
- Added the implementation for the oneof's `info`; let me know if you think it's ok 

Thanks a lot

tsaubergine commented:
Thanks - I'll take a deeper look later, but quick answers to your questions:

>  does the expected min/max size of a message include the header as well (i.e. the dccl.msg.id and the fields with in_head set to true)?

Yes, it includes the header as well, assuming you mean the call to dccl::Codec::size(): https://github.com/GobySoft/dccl/blob/3.0/src/codec.cpp#L429

I'm fine with excluding the use of oneof in the header, but I would like Codec to throw an exception if someone does mark a oneof field as `in_head: true`.

dfenucci commented:
Thanks for the feedback.

> I'm fine with excluding the use of oneof in the header, but I would like Codec to throw an exception if someone does mark a oneof field as `in_head: true`.

This is now implemented in [this commit](https://github.com/GobySoft/dccl/pull/70/commits/7d306b65002b246d6de6e9e12faff13f33d7f366)

tsaubergine commented:
Thanks, this looks good to me, now. I have [a bit more to do](https://github.com/GobySoft/dccl/projects/1) before I release DCCL4, but in the meantime you can access the DCCL4 pre-release Debian/Ubuntu packages as needed from the ["continuous" repository](https://goby.software/3.0/index.html#autotoc_md111):

```
dccl4-apps-dbgsym - debug symbols for dccl4-apps
dccl4-apps - Applications for the Dynamic Compact Control Language.
dccl4-compiler-dbgsym - debug symbols for dccl4-compiler
dccl4-compiler - Protobuf compiler for DCCL (protoc-gen-dccl): used to create Boost Units methods
dccl4-doc - Documentation for the Dynamic Compact Control Language.
libdccl4-dbgsym - debug symbols for libdccl4
libdccl4-dev - Developers' package for the Dynamic Compact Control Language.
libdccl4 - Libraries for the Dynamic Compact Control Language.
```

Thanks again for your valuable contribution.
